generator/internal/api/service_dependencies.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"maps"
+	"slices"
+)
+
+type ServiceDependencies struct {
+	Messages []string
+	Enums    []string
+}
+
+// FindServiceDependencies returns the message and enum IDs that are required by
+// a given service.
+//
+// The function traverses `model` starting from the definition of `serviceID`.
+//   - Any message used by a method of the service is included in the results.
+//   - Any message used by LROs of the service is included in the results.
+//   - The results are recursively scanned searching for any fields of the
+//     messages included above.
+//   - If a nested message is included in the results, then the parent message
+//     is also included (recursively) in the results.
+func FindServiceDependencies(model *API, serviceID string) *ServiceDependencies {
+	service, ok := model.State.ServiceByID[serviceID]
+	if !ok {
+		return &ServiceDependencies{}
+	}
+	state := newFindServicesState(serviceID, model)
+	state.seed(service)
+	state.search()
+	return &ServiceDependencies{
+		Messages: slices.Sorted(maps.Keys(state.Messages)),
+		Enums:    slices.Sorted(maps.Keys(state.Enums)),
+	}
+}
+
+type findServiceState struct {
+	// The message IDs that have already been visited
+	Visited map[string]bool
+	// The enums already included in these results
+	Enums map[string]bool
+	// The messages already included in these results
+	Messages   map[string]bool
+	model      *API
+	candidates []*Message
```

Comment:
do they not have a deque?

Reply:
> a stack works

yes. FTR, my comment was more about preferring container functions like `PopBack()` over manually implementing a stack using an array, so the code is more readable.